### PR TITLE
HDDS-10360. Make cleanupTest compatible with enableFileSystemPath values in TestKeyManagerImpl class.

### PR DESCRIPTION
**Pull request by ashishkumar50**
## What changes were proposed in this pull request?
When `enableFileSystemPath` is true and bucket layout is legacy, during key delete operation in normalizekey(prexecute) it removes trailing slash, but in DB key is present with trailing slash. And because of this KEY_NOT_FOUND error comes.
But fixing legacy bucket directory deletion can lead to other issues. Instead we are restricting legacy bucket directory deletion in [HDDS-10397](https://issues.apache.org/jira/browse/HDDS-10397). In this PR we are making cleanupTest compatible with enableFileSystemPath(true/false) values in TestKeyManagerImpl class.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10360

## How was this patch tested?

Existing test

### CODE REVIEW

**Comment by sadanand48** [comment]
Thanks @ashishkumar50 for the patch, In what cases does the DB store key with trailing slash here ? How was the key inserted ?  From what I see , if FS paths is enabled, even key creates are normalized and goes through same method (validateAndNormalizeKey) , see OmKeyCreateRequest so normalizing during delete is valid .

**Comment by ashishkumar50** [comment]
> ` [d18f2fe](/apache/ozone/pull/6218/commits/d18f2fee1dae7727d664de4392ef7765b80bf5b2) `

@sadanand48 Thanks for the review.
Assume when we create key like (Bucket is legacy with enableFileSystemPath)
ozone sh key put vol1/bucket1/xyz//def/key1 file1
It gets normalized and inserted into keyTable as (Parent trailing slash is preserved in OmKeyCreateRequest and that is correct):
1) xyz/
2) xyz/def/
3) xyz/def/key1

Now during delete 
1) xyz/def/key1 normalized as xyz/def/key1 and it is found in keyTable and deletion pass through
2) xyz/def/ gets normalized into xyz/def and since keyTable has xyz/def/ and so it is not found and throws KEY_NOT_FOUND error.

But we should preserve trailing slash here during delete.

**Comment by ashishkumar50** [comment]
> Thanks @ashishkumar50 for the patch, In what cases does the DB store key with trailing slash here ? How was the key inserted ? From what I see , if FS paths is enabled, even key creates are normalized and goes through same method (validateAndNormalizeKey) , see OmKeyCreateRequest so normalizing during delete is valid .

@sadanand48 Thanks for the review.
Assume when we create key like (Bucket is legacy with enableFileSystemPath)
ozone sh key put vol1/bucket1/xyz//def/key1 file1
It gets normalized and inserted into keyTable as (Parent trailing slash is preserved in OmKeyCreateRequest and that is correct):

xyz/
xyz/def/
xyz/def/key1
Now during delete

xyz/def/key1 normalized as xyz/def/key1 and it is found in keyTable and deletion pass through
xyz/def/ gets normalized into xyz/def and since keyTable has xyz/def/ and so it is not found and throws KEY_NOT_FOUND error.
But we should preserve trailing slash here during delete.

**Comment by sadanand48** [comment]
> Now during delete

How do we do this delete? Do we pass recursive flag here?  Do we have some commands to repro this?

**Comment by ashishkumar50** [comment]
> > Now during delete
> 
> How do we do this delete? Do we pass recursive flag here? Do we have some commands to repro this?

No recursive flag, just normal delete
```
ozone sh key put vol1/bucket1/xyz//def/key1 file1
bash-4.2$ ozone sh key delete vol1/bucket1/xyz/def/key1
bash-4.2$ ozone sh key delete vol1/bucket1/xyz/def/    
KEY_NOT_FOUND Key not found
bash-4.2$ ozone sh key delete vol1/bucket1/xyz/    
KEY_NOT_FOUND Key not found
```

**Comment by sadanand48** [comment]
One can argue here that `vol1/bucket1/xyz/def/` was not created by the user hence he should not be able to delete it, Although I'm not sure this change would break any other functionality, I think this issue has been always there and  not recently introduced. The key will be deleted when directory is deleted from FS interface.

**Comment by ashishkumar50** [comment]
> One can argue here that `vol1/bucket1/xyz/def/` was not created by the user hence he should not be able to delete it, Although I'm not sure this change would break any other functionality, I think this issue has been always there and not recently introduced. The key will be deleted when directory is deleted from FS interface.

Yes correct it seems issue is from either long time or from beginning, and there are some confusions as well.
When user executes `ozone sh key list vol1/bucket1` it shows intermediate directories as well.
But during delete it shows `KEY_NOT_FOUND` but it is listed in key list.

If we don't want to allow this delete, I think in that case it should fail in the client itself if user is trying to delete a key which is having trailing slash and show appropriate message as "Directory can be deleted only through FS interface".

**Comment by sadanand48** [comment]
> If we don't want to allow this delete, I think in that case it should fail in the client itself if user is trying to delete a key which is having trailing slash and show appropriate message as "Directory can be deleted only through FS interface".

+1 to this idea, intermediate dirs are created by the system to be compatible with the filesystem, they shouldn't be accessible as individual keys. I'm not aligned towards this particular change as  this is legacy code & it might break some other case which we are not aware of currently.

**Comment by ashishkumar50** [comment]
> > If we don't want to allow this delete, I think in that case it should fail in the client itself if user is trying to delete a key which is having trailing slash and show appropriate message as "Directory can be deleted only through FS interface".
> 
> +1 to this idea, intermediate dirs are created by the system to be compatible with the filesystem, they shouldn't be accessible as individual keys. I'm not aligned towards this particular change as this is legacy code & it might break some other case which we are not aware of currently.

Done this change to restrict legacy bucket directory deletion through sh command.

**Comment by ashishkumar50** [comment]
@sadanand48 , Shall I create another Jira for "restrict legacy bucket directory deletion through sh command." ?
In this Jira just make enableFileSystemPath as false in TestKeyManagerImpl class?

**Comment by sadanand48** [comment]
Sure @ashishkumar50

**Comment by ashishkumar50** [comment]
> @sadanand48 , Shall I create another Jira for "restrict legacy bucket directory deletion through sh command." ? In this Jira just make enableFileSystemPath as false in TestKeyManagerImpl class?

Created https://issues.apache.org/jira/browse/HDDS-10397

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java [hunk 1]
```diff
@@ -235,21 +241,21 @@ public static void cleanup() throws Exception {
...
-      if (fileStatus.isFile()) {
-        writeClient.deleteKey(
-            createKeyArgs(fileStatus.getKeyInfo().getKeyName()));
-      } else {
-        writeClient.deleteKey(createKeyArgs(OzoneFSUtils
-            .addTrailingSlashIfNeeded(
-                fileStatus.getKeyInfo().getKeyName())));
-      }
-    }
+    org.apache.hadoop.fs.Path volumePath = new org.apache.hadoop.fs.Path(OZONE_URI_DELIMITER, VOLUME_NAME);
```
Comment (sadanand48) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java:254]: import `org.apache.hadoop.fs.Path`

Comment (ashishkumar50) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestKeyManagerImpl.java:254]: There is already import of `import java.nio.file.Path;` so it's a conflict.

**Review by sadanand48 — APPROVED** [review]
LGTM

**Pull request merged**